### PR TITLE
Update API routes and validation schemas

### PR DESCRIPTION
**Pull request by gideonthomas**
Updated routes based on [this discussion](https://github.com/mozilla/publish.webmaker.org/issues/3)

### CODE REVIEW

File: routes/projects/files/create-update.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
+/*
+ * PUT /projects/:username/:projectName/files/:pathToFile
+ * The :pathToFile parameter must be URI encoded
+ * Create or replace the contents for the given file path
+ * The message body that would normally be expected by this route
+ * will look something like:
+ * {
+ *   type: <string> content-type,
+ *   path: <string>,
+ *   data: <string> base64 encoded binary data
```
Comment (humphd) [inline routes/projects/files/create-update.js]: It would be nice if this could be `String` (utf8) or `ArrayBuffer` (binary).  Having to deal in base64 for binary data sucks.

File: routes/projects/files/rename.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+ *   name: <string> required
+ * }
+*/
+
+var renameSchema = require('../../../schemas.js').rename;
+
+module.exports = function(server) {
+  server.route({
+    method: 'PUT',
+    path: '/projects/{username}/{projectName}/files/{path}/rename',
```
Comment (humphd) [inline routes/projects/files/rename.js]: Can one rename a file, and therefore `mv` it to a new path?  Or is rename always a change of name only for the leaf node?

File: routes/projects/files/get-one.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
+/*
+ * GET /projects/:username/:projectName/files/:pathToFile
+ * The :pathToFile parameter must be URI encoded
+ * Returns the contents of the given file in the form:
+ * {
+ *   type: <string> content-type,
+ *   path: <string>,
+ *   link: <string>
+ * }
```
Comment (humphd) [inline routes/projects/files/get-one.js]: I wonder if it's useful to also send `size` so that clients can judge the amount of work to download various resources (e.g., progress updates for opening a project).

File: routes/projects/update.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+ * PUT /projects/:username/:projectName
+ * Creates a new project for the given user. A project object is sent in
+ * the form:
+ * {
+ *   name: <string> required,
+ *   username: <string> required,
+ *   isPublic: <boolean> required,
+ *   publishedAt: <dataTime> optional,
+ *   publishLink: <string> optional,
+ *   files: [ <FileObj1>, <FileObj2>... ] required
```
Comment (humphd) [inline routes/projects/update.js]: Can this list of files be partial?

**Comment by ashleygwilliams** [comment]
looks like this would fix #8

File: routes/projects/files/rename.js (hunk 2 above)
Comment (sedge) [inline routes/projects/files/rename.js]: Rename is always a change of name, not the full path.

File: routes/projects/files/get-one.js (hunk 3 above)
Comment (sedge) [inline routes/projects/files/get-one.js]: +1

File: routes/projects/files/rename.js (hunk 2 above)
Comment (ashleygwilliams) [inline routes/projects/files/rename.js]: this is an interesting question, but i think we need to remember that a path and the name of a file are actually the same, e.g:

`/public/styles/custom.css` is the true name of `custom.css`

File: routes/projects/update.js (hunk 4 above)
Comment (sedge) [inline routes/projects/update.js]: Yes - further files should be added through `PUT /projects/:username/:projectname`

Comment (sedge) [inline routes/projects/update.js]: Though I wonder if this should be optional?

File: routes/projects/files/rename.js (hunk 2 above)
Comment (humphd) [inline routes/projects/files/rename.js]: So we'll have to do some work to maintain consistency across the backend storage view of a file vs. the filesystem, where the filesystem will only ever let us rename the leaf node.  It might be useful for us to disallow renaming `/public/styles/custom.css` to `/public/custom.css`, since it will lead to filesystem data loss.

File: routes/projects/files/create-update.js (hunk 1 above)
Comment (gideonthomas) [inline routes/projects/files/create-update.js]: For 'string' data, do you mean actual strings and not binary (encoded) data?

Comment (humphd) [inline routes/projects/files/create-update.js]: I'd suggest we always send/receive `ArrayBuffer` on the browser side, and always send/receive `Buffer` on the server side.  We don't need to encode at all.

Comment (gideonthomas) [inline routes/projects/files/create-update.js]: Actually, it's just my documentation that is wrong haha

File: routes/projects/update.js (hunk 4 above)
Comment (gideonthomas) [inline routes/projects/update.js]: It should basically represent whatever is in the user's fs (on the thimble side) at the moment this request is sent. So values can range between `[ ]` to `[ file1, file2, file3... file100 ]`

Comment (sedge) [inline routes/projects/update.js]: Even if those files aren't uploaded to the server? Doesn't a file object need to point at the file data?

Comment (gideonthomas) [inline routes/projects/update.js]: Hmm this does seem a bit problematic. @ashleygwilliams do you have any insight into this? What kind of foreign key validation do we want to do?

Comment (sedge) [inline routes/projects/update.js]: Since a foreign key won't exist yet, my vote is to keep it simple - Creating a project means it has no files associated with it yet. Adding files through the `file` routes can update the related `project` object.

Comment (gideonthomas) [inline routes/projects/update.js]: This is a PUT route though, which means updating project metadata will need you to provide all the list of files associated with the project...which can be problematic

File: routes/index.js [hunk 5]
```diff
@@ -4,25 +4,23 @@
...
+  require('./projects/files/delete.js')(server);
 
-  // /file/* API routes
-  // Handles project-specific file-level operations
-  require('./file/get-contents.js')(server);
-  require('./file/create-update.js')(server);
-  require('./file/delete.js')(server);
-  require('./file/rename.js')(server);
+  // RPC routes
+  require('./projects/rename.js')(server);
```
Comment (humphd) [inline routes/index.js]: Question: why do you separate the `*rename` routes out like this, rather than including with the associated routes above?

File: routes/projects/update.js (hunk 4 above)
Comment (sedge) [inline routes/projects/update.js]: If we consider "files" only to exist if they exist on the server then this should be okay.

File: routes/index.js (hunk 5 above)
Comment (sedge) [inline routes/index.js]: This was because they aren't REST endpoints, since they're fire and forget operations. They're meant to condense the process of renaming down to one call.

Comment (gideonthomas) [inline routes/index.js]: Hmm I guess I thought it fit well with the type of route it is. But it probably makes more sense to associate it with the data it is working on i.e. project or file

Comment (gideonthomas) [inline routes/index.js]: Well it honestly is a matter of preference, so either way works

File: routes/projects/update.js (hunk 4 above)
Comment (gideonthomas) [inline routes/projects/update.js]: well that means that the client needs to make sure that we only send files that are on the server and not "queued up for autosave"

File: routes/projects/create.js [hunk 6]
```diff
@@ -0,0 +1,28 @@
+/*
+ * POST /projects
+ * Creates a new project for the given user. A project object is sent in
+ * the form:
+ * {
+ *   name: <string> required,
+ *   username: <string> required,
+ *   isPublic: <boolean> required,
+ *   publishedAt: <dataTime> optional,
+ *   publishLink: <string> optional,
```
Comment (humphd) [inline routes/projects/create.js]: The client won't send `publishedAt` or `publishLink`, so I think this is confusing.  It will get *stored* with this data, but I don't think you want to require it to be *sent* in this form.

I think there are optional bits to this.  For example:

```js
{
  name: "...",
  username: "...",
  files: [initial, list, of, files]
}
```

In this case, `isPublic` can take the value `false` since it is absent, and `publishedAt` and `publishLink` will get updated later.

File: routes/projects/update.js (hunk 4 above)
Comment (sedge) [inline routes/projects/update.js]: How so? AFAIK it just means that we populate the project object with file references as we upload them to the server for the first time.

File: routes/projects/delete.js [hunk 7]
```diff
@@ -1,12 +1,12 @@
 /*
- * DELETE /project/:username/:projectName
+ * DELETE /projects/:username/:projectName
  * Deletes the given project for the given user
```
Comment (humphd) [inline routes/projects/delete.js]: I assume this cascades to also delete all files associated with the project, as well as published resources?  If so, call it out here.

File: routes/projects/files/create-update.js [hunk 8]
```diff
@@ -0,0 +1,28 @@
+/*
+ * PUT /projects/:username/:projectName/files/:pathToFile
+ * The :pathToFile parameter must be URI encoded
+ * Create or replace the contents for the given file path
+ * The message body that would normally be expected by this route
+ * will look something like:
+ * {
+ *   type: <string> content-type required,
+ *   path: <string> required,
+ *   size: <integer> required,
```
Comment (humphd) [inline routes/projects/files/create-update.js]: This isn't necessary, since it can be determined by the size of the body contents.  I'd remove it.

File: routes/projects/update.js (hunk 4 above)
Comment (sedge) [inline routes/projects/update.js]: A sequence diagram would help here -.-

1. Create a project with POST /projects
2. Upload existing files for that with PUT /projects/:username/:projectname/files/:filename, which adds them as foreign keys in the project object

Comment (ashleygwilliams) [inline routes/projects/update.js]: ok- so a few thoughts:

when a project is created, i assume that a user will be redirected to a view that has a file open and ready to edit (likely `index.html`). this default file should be created at the time the project is created. that way a foreign key will exist. all projects, to be projects need at least one file, IMO.

thoughts?

File: routes/projects/files/get-one.js [hunk 9]
```diff
@@ -0,0 +1,19 @@
+/*
+ * GET /projects/:username/:projectName/files/:pathToFile
+ * The :pathToFile parameter must be URI encoded
+ * Returns the contents of the given file in the form:
+ * {
+ *   type: <string> content-type,
+ *   path: <string>,
+ *   size: <integer>,
+ *   link: <string>
```
Comment (humphd) [inline routes/projects/files/get-one.js]: What is `link`?  Where is the actual `Buffer` for this file?

File: routes/projects/files/create-update.js (hunk 8 above)
Comment (gideonthomas) [inline routes/projects/files/create-update.js]: as in make it optional?

File: routes/projects/files/rename.js (hunk 2 above)
Comment (humphd) [inline routes/projects/files/rename.js]: @gideonthomas I think you need to update the comment above to indicate that we only allow `filename` changes, not `pathname`.

File: routes/projects/files/get-one.js (hunk 9 above)
Comment (gideonthomas) [inline routes/projects/files/get-one.js]: I should probably add a little more documentation here. I think our plan was to provide a `link` to the file contents (maybe in s3) as opposed to adding it to the metadata
cc @ashleygwilliams

File: routes/projects/update.js (hunk 4 above)
Comment (ashleygwilliams) [inline routes/projects/update.js]: re: updating files- new files should be created at `POST /files`, not at `PUT /projectname`. though we could implement a `PATCH /projectname` but that gets gross since the project should really only contain the foreign keys, not the actual file object representation...

tl;dr files should be made and updated at `POST/PUT /files`. they'll be associated with the project by foreign key. so when the `files` attribute is passed in a `PUT /projectname` it shouldn't be updating anything.

Comment (humphd) [inline routes/projects/update.js]: I think this API's view of files associated with a project has to relate to what's on the server, which is to say, I don't think a client should be able to specify it, only create, delete, and rename files for a project.  Since this payload won't include those files, I'd suggest you remove this, and just use `[]` until the files themselves are uploaded.

Comment (sedge) [inline routes/projects/update.js]: @ashleygwilliams Hmm. Projects will be populated by default, but only on the client side since they won't exist as "project objects" until `POST /projects` is hit. At that point, the files haven't been uploaded to the server yet, so I don't see how there can be reference to them.

Unless we want to create a new "project object" every time a user opens thimble. With a generic name maybe?

Comment (sedge) [inline routes/projects/update.js]: @humphd +1

File: routes/projects/files/get-one.js (hunk 9 above)
Comment (ashleygwilliams) [inline routes/projects/files/get-one.js]: yes that is correct. i think serving a blob from the API is not a great idea. passing a representation of the file and not the file itself seems best in my opinion. the blob should be served statically.

Comment (sedge) [inline routes/projects/files/get-one.js]: It looks like we're leaning towards postgres. I think it would be simplest to just return the buffer here.

File: routes/projects/update.js (hunk 4 above)
Comment (gideonthomas) [inline routes/projects/update.js]: So what's the verdict on this?
POST /projects body:
```js
{
  ...
  files: []
}
```
PUT /projects/:username/:prjName
```
same as above
```
GET /projects/:username/:prjName
```
{
  ...
  files: [ 
            { 
               ..., 
               link: linkTofile 
            }, 
            ... 
         ]
}
```

Is that it?
@ashleygwilliams @humphd @sedge

Comment (sedge) [inline routes/projects/update.js]: We need to figure out how we're going to access the data for a file before we can definitively answer this question.

File: routes/projects/delete.js (hunk 7 above)
Comment (sedge) [inline routes/projects/delete.js]: Yeah it would.

File: schemas.js [hunk 10]
```diff
@@ -0,0 +1,25 @@
+/*
+ * Validation schemas for incoming requests
+*/
+
+var Joi = require('joi');
+var mimes = Object.keys(require('mime-db'));
+
+exports.file = Joi.object().keys({
+  type: Joi.string().valid(mimes),
```
Comment (sedge) [inline schemas.js]: Don't we have restrictions on which files we'll accept into a project? This would be a good place to enforce that.

Comment (gideonthomas) [inline schemas.js]: Are there restrictions on which files we'll accept? Anyway, this is not going to be taken in the PUT/POST so the client doesn't really have control over this, the server determines what the mime type is.

Comment (ashleygwilliams) [inline schemas.js]: IMO that would be an enhancement. let's not worry about that for now, ship a simple version of this and move the thread to the issues.

Comment (sedge) [inline schemas.js]: +1
